metrique-aggregation/src/histogram.rs

```diff
@@ -0,0 +1,439 @@
+
+use metrique_core::CloseValue;
+use metrique_writer::{MetricFlags, MetricValue, Observation, Value, ValueWriter};
+use smallvec::SmallVec;
+use std::marker::PhantomData;
+
+/// Strategy for aggregating observations in a histogram.
+///
+/// Implementations determine how values are stored and converted to observations
+/// when the histogram is closed.
+pub trait AggregationStrategy {
+    /// Record a single observation.
+    fn record(&mut self, value: f64);
+
+    /// Drain all observations and return them as a vector.
+    ///
+    /// This resets the strategy's internal state.
+    fn drain(&mut self) -> Vec<Observation>;
+}
+
+/// Thread-safe strategy for aggregating observations in a histogram.
+///
+/// Like [`AggregationStrategy`] but allows recording values through a shared reference.
+pub trait AtomicAggregationStrategy {
+    /// Record a single observation through a shared reference.
+    fn record(&self, value: f64);
+
+    /// Drain all observations and return them as a vector.
+    ///
+    /// This resets the strategy's internal state.
+    fn drain(&self) -> Vec<Observation>;
+}
+
+/// A histogram that collects multiple observations and emits them as a distribution.
+///
+/// Use this when you have many observations of the same metric within a single unit of work.
+/// The histogram aggregates values in memory and emits them as a single metric entry.
+///
+/// Requires `&mut self` to add values. For thread-safe access, use [`AtomicHistogram`].
+pub struct Histogram<T, S = ExponentialAggregationStrategy> {
+    strategy: S,
+    _value: PhantomData<T>,
+}
+
+impl<T, S: AggregationStrategy> Histogram<T, S> {
+    /// Create a new histogram with the given aggregation strategy.
+    pub fn new(strategy: S) -> Self {
+        Self {
+            strategy,
+            _value: PhantomData,
+        }
+    }
+
+    /// Add a value to the histogram.
+    ///
+    /// The value is converted to observations using the metric value's implementation,
+    /// then recorded in the aggregation strategy.
+    pub fn add_value(&mut self, value: T)
+    where
+        T: MetricValue,
+    {
+        struct Capturer<'a, S>(&'a mut S);
+        impl<'b, S: AggregationStrategy> ValueWriter for Capturer<'b, S> {
+            fn string(self, _value: &str) {}
+            fn metric<'a>(
+                self,
+                distribution: impl IntoIterator<Item = Observation>,
+                _unit: metrique_writer::Unit,
+                _dimensions: impl IntoIterator<Item = (&'a str, &'a str)>,
+                _flags: MetricFlags<'_>,
+            ) {
+                for obs in distribution {
+                    match obs {
+                        Observation::Unsigned(v) => self.0.record(v as f64),
+                        Observation::Floating(v) => self.0.record(v),
+                        Observation::Repeated { total, occurrences } => {
+                            let avg = total / occurrences as f64;
```

Comment:
We should have some sort of record_many there - occurrences can be large.

It is fine if the sorted aggregation strategy does not have a count field, but the exponential aggregation strategy should be O(1) in the number of occurrences.

Reply:
you should handle occurrences == 0